### PR TITLE
wip: sketch out binstrings

### PR DESCRIPTION
**Pull request by leostera**
Opening this PR to start sketching what a binary string implementation would look like.

Ideally we'd want:

- [x] O(1) concats
- [x] O(1) views 
- [ ] UTF-8 capable, so we can extract valid utf8/16 graphemes from it
- [ ] conversions from/to String

And we'll need a few things to support compiling the patterns down to ops over the bin string:

- [ ] get a byte - to support `{| 1,x,3 |} = "123" -> x = 2`
- [ ] get a byte-range - to support `{| x::bytes(2), 3 |} = "123" -> x = "12"`
- [ ] get a specific bit – to support `hello::1`
- [ ] get the tail of a binstring - to support `{| head, tail::binary |}` 
- [ ] get a utf8 characters - to support `c::utf8`

A good test-case would be to parse a WebSocket text frame and get the first utf8 off the payload, without copying _any_ data off the string.

```ocaml
let frame {%bin| fin::1, comp::1, _rsv::2, 1::4, 0::1, 127::7, len::64, mask::32, payload::bytes(len) |} =
  let%bin {| head::utf8, rest::binary |} = payload in
  Printf.printf "%s" head
```

### CODE REVIEW

File: binstring/binstring.ml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+exception Malformed of string
+
+let of_string str =
+  let length = Str.length str in
+  let buf = Bytes.create length in
+  let decoder = Uutf.decoder ~encoding:`UTF_8 (`String str) in
+  let pos = ref 0 in
+  let add char =
+    let n = Bytes.set_utf_8_uchar buf !pos char in
```
Comment (felipecrv) [inline binstring/binstring.ml:17]: Is this setting an UTF-8 at any random position? That sounds expensive.

File: binstring/binstring.ml [hunk 2]
```diff
@@ -0,0 +1,125 @@
+module Str = Stdlib.String
+module Buf = Stdlib.Buffer
+
+type t = { inner : string list; offset : int; length : int }
```
Comment (felipecrv) [inline binstring/binstring.ml:4]: This internal representation can change later, right?

Comment (leostera) [reply binstring/binstring.ml:4]: 100% – this is an abstract type from the outside.

File: binstring/binstring.ml (hunk 1 above)
Comment (leostera) [reply binstring/binstring.ml:17]: we have a preallocated buffer here with the size of the string that we're verifying for utf-8 encoding, so this is more of an array.set.

in any case, after our chat, i'm removing this verification and leaving it to the extraction functions.

File: binstring/binstring.ml (hunk 2 above)
Comment (leostera) [reply binstring/binstring.ml:4]: done!

**Review by felipecrv — APPROVED** [review]

**Comment by leostera** [comment]
So after exploring using these strings on stream today we found that we want to expand the IO.Reader/Writer interfaces to allow for easy vectored reading/writing.

Next steps to prepare for using these within Riot:

* Add `write_owned_vectored` to the `Write` interface
* Implement `to_writer` for Bytestring, using a Transient to build the bytestring, where `flush` is a call to `Transient.commit`
* Implement `Bytestring.with_writer (fun writer -> ...)` as a shorthand to create throwaway writers
* Add `read_shared_vectored` to the `Read` interface
* Implement `to_reader` for Bytestring

**Comment by leostera** [comment]
Okay all those changes are in place, we should be good to do a Bytestring sink now.

**Pull request merged**